### ISSUE TITLE
multitenant: bogus row in `system.tenant_settings` may prevent tenant upgrade

### ISSUE THREAD

**Issue by renatolabs** [cockroachdb/cockroach#125702]
**TLDR**:

Clusters that upgraded to 23.1 using a version older than v23.1.9 will forever have a row in the `system.tenant_settings` table for `tenant_id=0` and `version=23.1`. This may prevent tenants from upgrading in the future as they might have an incorrect view of the storage cluster's version in specific circumstances.

*****

### Long winded background

In the 23.1 release cycle, we introduced a startup migration[^1] (later rewritten as a permanent upgrade) that added a row to the `system.tenant_settings` table with the storage cluster's version information. Tenants would get notified of changes in this table (via the `settingswatcher`) in order to keep an updated view of the storage cluster's version.

However, the startup migration had a bug in its implementation; specifically, it used the `%d` 'verb' to generate the query, but the `tenantID` variable is of type `*int64`[^2]:

https://github.com/cockroachdb/cockroach/blob/fd42d376fb5ec42e9d9a30fcdcb210a8a79ef0e2/pkg/startupmigrations/migrations.go#L760-L770

Similar logic was added when changing the `version` setting in the system tenant; the version would be updated accordingly in this new row in `system.tenant_settings`:

https://github.com/cockroachdb/cockroach/blob/fd42d376fb5ec42e9d9a30fcdcb210a8a79ef0e2/pkg/sql/set_cluster_setting.go#L558-L569

Note, however, that _this_ implementation doesn't have same bug: while we still use `tree.NewDInt(0)`, we're using the `$1` placeholder instead of `fmt.Sprintf`.

The bug with the permanent upgrade was eventually noticed (#105858) and fixed (#105854). Shortly after, the way tenants learn about changes in the storage cluster's version also changed to no longer rely on an entry in the `system.tenant_settings` table: #106037. Both of these fixes were backported to the 23.1 release branch and were available in public releases starting from the v23.1.9 version.

A corollary to all of this is that the state of the `system.tenant_settings` table is **different depending on whether the cluster was bootstrapped on 23.1 vs bootstrapped on an older version and upgraded to 23.1**.

<details>
<summary>Bootstrapping on 23.1</summary>

```console
$ roachprod create local -n1
$ roachprod stage local release v23.1.8
$ roachprod start local
$ roachprod sql local:1 -- -e "SELECT * FROM system.tenant_settings;"
    tenant_id   |  name   |               value               |        last_updated        | value_type | reason
----------------+---------+-----------------------------------+----------------------------+------------+---------
  1374467842288 | version | \x12\b\b\x17\x10\x01\x18\x00 \x00 | 2024-06-14 18:37:22.956285 | m          | NULL
(1 row)
```
</details>

<details>
<summary>Upgrading to 23.1</summary>

```console
$ roachprod create local -n1
$ roachprod stage local release v22.2.19
$ roachprod start local
$ roachprod deploy local release v23.1.8
$ roachprod sql local:1 -- -e "SET CLUSTER SETTING version = '23.1';" # wait for migrations
$ roachprod sql local:1 -- -e "SELECT * FROM system.tenant_settings;"
  tenant_id |  name   |               value               |        last_updated        | value_type | reason
------------+---------+-----------------------------------+----------------------------+------------+---------
          0 | version | \x12\b\b\x17\x10\x01\x18\x00 \x00 | 2024-06-14 18:43:23.225887 | m          | NULL
(1 row)
```
</details>

As can be seen, the `tenant_id` column is `0` only when upgrading go 23.1; this is relevant because `0` encodes "all-tenants" overrides[^3].

### The bug

The fix introduced in #106037 (and present in v23.1.9+) works around potentially wrong entries in the `system.tenant_settings` table by "injecting" the real storage cluster version when starting up:

https://github.com/cockroachdb/cockroach/blob/a62d88b12af22f8eeecdae1f0f095777241e47e8/pkg/server/node.go#L2071-L2077

This works reasonably well; however, there are two other situations where the tenant's view of cluster settings (including the storage cluster's version) change (or rather, is overwritten):

* when a `SystemVisible` cluster setting is updated[^4]
* when a cluster setting is updated for _all tenants_[^5]

Since, in these cases, the `version` entry is not being overwritten, the tenant will update its local cache with whatever is present in the `system.tenant_settings` table (which will be outdated if the cluster has upgraded from 23.1 since).

It's easy to reproduce this situation with roachprod (note that the script sets up a shared-process tenant but the same issue exists for separate-process tenants):

<details>
<summary>Reproduction</summary>

```console
$ roachprod create local -n1
$ roachprod stage local release v22.2.19
$ roachprod start local
$ roachprod deploy local release v23.1.8
$ roachprod sql local:1 -- -e "SET CLUSTER SETTING version = '23.1';" # wait for migrations
$ roachprod start-sql my-tenant --storage-cluster local # create shared-process tenant
$ roachprod deploy local release v23.2.5 # upgrade to 23.2
$ roachprod sql local:1 -- -e "SET CLUSTER SETTING version = '23.2';" # wait for migrations
$ roachprod sql local:1 -- -e "SET CLUSTER SETTING kv.closed_timestamp.target_duration = '5s';" # set an arbitrary SystemVisible setting
$ roachprod sql local:1 --cluster my-tenant -- -e "SET CLUSTER SETTING version = '23.2';" # attempt to upgrade tenant
```

The last command in that script fails with:

```
ERROR: preventing tenant upgrade from running as the storage cluster has not yet been upgraded: storage cluster version = 23.1, tenant cluster version = 23.1
```

which is wrong since we just upgraded the cluster to 23.2 two commands above.

Note: this uses the "updating a `SystemVisible` setting" approach to reproducing the bug. We could also change the default for all tenants with something like `"ALTER VIRTUAL CLUSTER ALL SET CLUSTER SETTING {ApplicationLevel setting};"` and the result would be the same.
</details>

### Possible solutions

I can imagine at least a few ways we could go about dealing with this situation:

* Write a migration to delete the `system.tenant_settings` entry where `tenant_id=0` and `name=version`. To me this would be ideal as it would definitely clear up the bogus row in that table. The fact that this wasn't done in the 23.2 cycle may suggest this could have undesired consequences.

* Apply the same "injection" of the storage cluster version done at startup in the two situations mentioned above (changing `SystemVisible` and "all-tenants" defaults). Not as clean, and the bug could be reintroduced if we add some other way to refresh the tenant's view of these settings.

* Reject regressions in the storage cluster version in the code below, as that would indicate a programming error. This is also not ideal as it could mask other issues:

https://github.com/cockroachdb/cockroach/blob/1ed54694b792af3d9c90961fb32e690adb2a1325/pkg/server/settingswatcher/settings_watcher.go#L514-L515

* Do nothing and if anyone finds themselves in this situation, ask them to restart the nodes, as that will cause the startup logic to kick in and the right storage cluster version will be cached. Unsatisfying IMO.

[^1]: see #91071 for details.
[^2]: https://github.com/cockroachdb/cockroach/blob/fd42d376fb5ec42e9d9a30fcdcb210a8a79ef0e2/pkg/sql/sem/tree/datum.go#L689-L691
[^3]: for definition and uses, see: https://github.com/cockroachdb/cockroach/blob/1ed54694b792af3d9c90961fb32e690adb2a1325/pkg/server/tenantsettingswatcher/row_decoder.go#L34
[^4]: https://github.com/cockroachdb/cockroach/blob/1ed54694b792af3d9c90961fb32e690adb2a1325/pkg/server/tenantsettingswatcher/watcher.go#L270-L282
[^5]: https://github.com/cockroachdb/cockroach/blob/1ed54694b792af3d9c90961fb32e690adb2a1325/pkg/server/node.go#L2459-L2466

Jira issue: CRDB-39563

**Comment by renatolabs** [comment]
Ownership of any of this is undefined, so I'm pinging some folks involved with this work when it happened and former members of the V-team: @ajstorm, @yuzefovich, @stevendanna.

Apologies for the long writeup, but I figured it would be helpful to include more context as the details of all this could be less clear in your memories by now.

I'm happy to work on some fix for this, but would appreciate your input as to the preferred direction to take.

**Comment by stevendanna** [comment]
Thanks for the write up.

Another interesting note is that we still seem to allow writing an override for the version key via `ALTER TENANT .. SET CLUSTER SETTING`.  I suppose that could be useful in some situation where we want to hide the real storage cluster version form a tenant for some reason, but it still seems like a footgun to me.

> Write a migration to delete the system.tenant_settings entry where tenant_id=0 and name=version. To me this would be ideal as it would definitely clear up the bogus row in that table. The fact that this wasn't done in the 23.2 cycle may suggest this could have undesired consequences.

I'm having trouble seeing a downside.  My guess is that this wasn't done because the assumption was that https://github.com/cockroachdb/cockroach/pull/106037 fixed the problem.

> Apply the same "injection" of the storage cluster version done at startup in the two situations mentioned above (changing SystemVisible and "all-tenants" defaults). Not as clean, and the bug could be reintroduced if we add some other way to refresh the tenant's view of these settings.

I think we should _also_ do this if we do a migration as this can be backported.  So this fix will prevent a problem for anyone upgrading to 24.1 and then a 24.2 migration will remove the bad data.

**Comment by ajstorm** [comment]
> I'm having trouble seeing a downside. 

I too don't see a downside. I agree with this approach.
